vortex-layout/src/layouts/row_id.rs

```diff
@@ -0,0 +1,366 @@
+                // Partition the expression into expressions that can be evaluated over the row_id field
+                // and all other fields that a delegated to their children.
+                Arc::new(
+                    var_partitions_with_map(expr, |id| {
+                        if *id == *ROW_ID {
+                            ROW_ID.clone()
+                        } else {
+                            "".into()
+                        }
+                    })
+                    .vortex_expect("We should not fail to partition variables"),
+                )
+            })
+            .clone()
+    }
+}
+
+impl LayoutReader for RowIdLayoutReader {
+    fn name(&self) -> &Arc<str> {
+        &self.name
+    }
+
+    fn dtype(&self) -> &DType {
+        self.child.dtype()
+    }
+
+    fn row_count(&self) -> Precision<u64> {
+        self.child.row_count()
+    }
+
+    fn register_splits(
+        &self,
+        field_mask: &[FieldMask],
+        row_offset: u64,
+        splits: &mut BTreeSet<u64>,
+    ) -> VortexResult<()> {
+        self.child.register_splits(field_mask, row_offset, splits)
+    }
+
+    fn pruning_evaluation(
+        &self,
+        row_range: &Range<u64>,
+        expr: &ExprRef,
+    ) -> VortexResult<Box<dyn PruningEvaluation>> {
+        // TODO(joe): add variable and then row_id to the pruning eval
+        self.child.pruning_evaluation(row_range, expr)
+    }
+
+    fn filter_evaluation(
+        &self,
+        row_range: &Range<u64>,
+        expr: &ExprRef,
+    ) -> VortexResult<Box<dyn MaskEvaluation>> {
+        let arr_len = row_range.clone().count();
+        let partitioned = self.partition_expr(expr);
+        let Some(row_id) = partitioned.find_partition(&ROW_ID) else {
+            return self.child.filter_evaluation(row_range, expr);
+        };
+        let rest = partitioned.find_partition(&Arc::from(IDENTITY_IDENTIFIER));
+
+        let row_id_scope = Scope::empty(arr_len).with_value(
+            ROW_ID.clone(),
+            SequenceArray::typed_new(row_range.start, 1, arr_len)
+                .vortex_expect("cannot be out of bounds")
+                .to_array(),
+        );
+
+        let rest_eval = if let Some(rest) = &rest {
+            let dtype = rest.return_dtype(&ScopeDType::new(self.child.dtype().clone()))?;
+            Some(if matches!(dtype, DType::Bool(_)) {
+                // If the partition evaluates to a boolean, we can evaluate it as a mask which
+                // can often be more efficient since nulls are turned into `false` early on,
+                // and layouts can perform predicate pruning / indexing.
+                FieldEval::Mask(self.child.filter_evaluation(row_range, expr)?)
+            } else {
+                // Otherwise, we evaluate the projection as an array, and combine the results
+                // at the end.
+                FieldEval::Array(self.child.projection_evaluation(row_range, expr)?)
+            })
```

Comment:
A follow up but having this both on struct reader and here I wonder if this warrants to be more generic

Reply:
I think that this should disappear, I have a PR https://github.com/vortex-data/vortex/pull/3487